keps/sig-node/20190626-node-shutdown-state.md

```diff
@@ -0,0 +1,253 @@
+- Apply the node-shutdown taint
+- removes forcefully the pods
+- Removes the taint
+
+given the current design of our controllers and their following of the single-responsibility pattern, we’ll introduce a new controller called gc_controller under pkg/controller/cloud/podgc.
+
+This controller assumes running on the out-of-tree CCM, to avoid piling features on the deprecated in-tree providers.
+
+This cloud gc_controller watches node objects and when notified (each step describe the possible race condition that could happen)
+
+1. checks the node state via IsShutdownByProviderID(): if there’s a race condition with the cloud provider (i.e. the instance came back see 2.a)
+2. apply a node-shutdown taint
+3. tries to hold node heartbeat lease (Locking is client-side relying on the holder and resourceVersion fields)
+  3.a. fails: backoff and retry if max retries hit -> aborts
+  3.b. success: hold the lease for this node: if the node comes back at this moment it won’t be able to acquire the lease -> pods won’t start
+4. register node as being processed in order to renew the Lease periodically
+5. delete the pod with a 0-grace-period
+7. once 5. returns untaint the node and stop renewing the Lease
+
+
+To avoid being too reactive in 2. we can backoff and not try to acquire the lock immediately to avoid evicting too fast
+
+#### Kubelet
+
+we need also to apply changes to the kubelet, at startup the kubelet will:
+
+1. acquire the Lease:
+  a. fails: do not start any pods, unless the pod tolerates the node shutdown taint
+  b. success: start normally
+
+### Risks and Mitigations
+
+This KEP introduces changes, and if not tested correctly could result in data corruption for users.
+To mitigate this we plan to have a high test coverage and to introduce this enhancement as an alpha
+
+## Design Details
+
+### Test Plan
+
+This feature will be tested with the following approaches:
+
+- unit tests
+- integration
+- e2e
+
+we also plan to test this with different version Skews.
+
+### Graduation Criteria
+
+This KEP will be treated as a new feature, and will be introduced with a new feature gate, NodeShutdownFailover.
+
+This enhancement will go through the following maturity levels: alpha, beta and stable.
+
+Graduation criteria between these levels to be determined.
+
+### Upgrade / Downgrade Strategy
+
+If applicable, how will the component be upgraded and downgraded? Make sure this is in the test plan.
+
+Consider the following in developing an upgrade/downgrade strategy for this enhancement:
+- to keep the old behavior: disable the feature gate
+- make sure that the new behavior is not running against lower-version Kubelets
+
+### Version Skew Strategy
+
+This feature requires version parity between the control plane and the kubelet, this behavior shouldn't be enabled in case of older Kubelets
+or Kubelets disabling the feature gate. To do this, we will make the kubelet add annotation `lease.kubernetes.io/node-failover=true` 
```

Comment:
If a kubelet is downgraded, how will the annotation be removed?

Reply:
good point. I'll rollback to my previous previous proposition, which was a flag, maybe something generic like `--honor-taints`, thoughts @dashpole ?

Reply:
What would the flag do?  I'm not familiar with the previous proposition.

Reply:
@dashpole @misterikkit - thinking more about this, this wouldn't be needed as the Lease object constructed by a kubelet on n-1 version would be without the annotation, so this wouldn't be a problem. I think we can keep the annotation.

Reply:
seems like we need to introduce a change in 1.16 that checks and removes the annotation so we can downgrade, thoughts @liggitt ?